_rules/html-page-lang-matches-default-ucwvc8.md

```diff
@@ -0,0 +1,227 @@
+
+## Expectation
+
+For each test target, the [primary language][] of the [valid language tag][] matches the [default page language][] of the test target.
+
+## Assumptions
+
+- The way to determine the default language of a page should be done by counting the number of visible words used in each language. If the default language needs to be derived in some other way, this rule may produce incorrect results.
+
+- The language of the page can be set by other methods than the `lang` attribute, for example using HTTP headers or the `meta` element. These methods are not supported by all assistive technologies. This rule assumes that these other methods are insufficient to satisfying [Success Criterion 3.1.1: Language of Page](https://www.w3.org/TR/WCAG21/#language-of-page).
+
+- This rule assumes that user agents and assistive technologies can programmatically determine [valid language tags](#valid-language-tag) even if these do not conform to the [BCP 47][] syntax.
+
+- This rule assumes that [grandfathered tags][] are not used as these will not be recognized as [valid language tags](#valid-language-tag).
+
+## Accessibility Support
+
+_There are no major accessibility support issues known for this rule._
+
+## Background
+
+- [Understanding Success Criterion 3.1.1: Language of Page](https://www.w3.org/WAI/WCAG21/Understanding/language-of-page.html)
+- [H57: Using language attributes on the html element](https://www.w3.org/WAI/WCAG21/Techniques/html/H57)
+- [BCP 47: Tags for Identifying Languages](https://www.ietf.org/rfc/bcp/bcp47.txt)
+- [The `lang` and `xml:lang` attributes](https://html.spec.whatwg.org/multipage/dom.html#the-lang-and-xml:lang-attributes)
+
+## Test Cases
+
+### Passed
+
+#### Passed Example 1
+
+This page has a [default language][default page language] of English, which matches the language indicated in the `lang` attribute.
+
+```html
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>ACT Rules Format 1.0 - Abstract</title>
+	</head>
+	<body>
+		<p>
+			The Accessibility Conformance Testing (ACT) Rules Format 1.0 defines a format for writing accessibility test
+			rules. These test rules can be used for developing automated testing tools and manual testing methodologies. It
+			provides a common format that allows any party involved in accessibility testing to document and share their
+			testing procedures in a robust and understandable manner. This enables transparency and harmonization of testing
+			methods, including methods implemented by accessibility test tools.
+		</p>
+	</body>
+</html>
+```
+
+#### Passed Example 2
+
+This page has a [default language][default page language] of English because most words are English. The [default page language][] matches the language indicated in the `lang` attribute.
+
+```html
+<!DOCTYPE html>
+<html lang="en">
+	<head>
+		<title>Gelukkig</title>
+	</head>
+	<body>
+		<p>The Dutch word "gelukkig" has no equivalent in English.</p>
```

Comment:
I am a bit worried about examples like that (due to the definition of default language).

`The word ale is from "øl"`
Both "the" (alternate spelling of the correct "te"), "ale", "is" and "from" are also Danish words (according to ordnet.dk). So this sentence has actually 5 English words and 5 Danish words, thus no default language. Although it has absolutely no meaning as a sentence in Danish.

Now, this is OK-ish, there is no default language so we don't do anything.

What I'm worried of is that we can tweak such an example enough to have a sentence with, say, 8 words that are English and 10 that are Danish, but just because some English words happen to have a completely different meaning in Danish. The sentence would be undoubtfully English, but the default language would be Danish.

Then, I guess this would be a case where the first assumption is wrong, so that is correct…

Maybe strengthen the first assumption by giving examples of other way to detect language? (letters/words frequency analysis, mutual information based distance, n-grams, …)

Reply:
I thought of that too. This seems fairly hypothetical though. Real world web pages don't just have a single phrase on them. I'm counting on pages having enough content on them to make things like the same word existing in multiple language not matter.

There are a number of other problems with this approach. The biggest one I have with it is that it's difficult to do accurately. If a page is half English half Danish, you can obviously word count the thing, but grabbing all the "Danish" words from the English section, is difficult to do. I explored some other ideas for this, like looking at sentences or paragraphs and counting those, but there you get into issues of how to define those. Also how do you compare two short paragraphs or two short sentences to a single long one. There's also the issue that some languages concatenate words. The same paragraph in German will have fewer words in it than the English one, just because in English words aren't concatenated.


The reason I went with this approach is because it is going to be easy to do for 99.9% of all pages. You can pretty much just eyeball it in almost every possible case, and in the cases they can't, a simple word count tool will resolve the problem. That extreme case where it is too close to call, yes this isn't ideal, but solving for that will probably take months, and is likely going to add complexity to testing this for the 99.9%. I don't think that's worth it.